sudoku_solver/draft_one/lib/doubly_linked_list.rb

```diff
@@ -0,0 +1,75 @@
+module DoublyLinkedList
```

Comment:
My initial reaction to this was "It's strange that this is a module."  I caught myself and went to double-check how you were using it, but I stand by my initial reaction. :smile: 

Both `DoublyLinkedList` and `NTree` should be their own classes.  Other classes can make use of them internally (like the `Sudoku::Puzzle` class below), but they should be carrying around their own internal copies.  See the Wikipedia article on [Composition Over Inheritance](http://en.wikipedia.org/wiki/Composition_over_inheritance).

Doing it this way makes for much more confusing (and tightly coupled) code.  There's a huge amount of indirection, i.e., it's not at all clear whether a particular method is one defined on this class or one coming from one of the included modules.  Likewise, your `Sudoku::Puzzle` class knows all sorts of details about the inner workings of the `DoublyLinkedList` module, reflected in the direct manipulation of instance variables like `@head`, `@tail`, and so on.

Think about how the [Enumerable](http://ruby-doc.org/core-2.2.0/Enumerable.html) or [Comparable](http://www.ruby-doc.org/core-2.2.0/Comparable.html) modules work in Ruby.  In both cases, including the module adds a bunch of behavior to the class into which you included it.  However, the underlying module code requires _only_ that the including class implement one specific method to get that behavior (`each` in the case of `Enumerable` and `<=>` in the case of `Comparable`).  The underlying module implements everything in terms of that single method and possibly other module-defined methods.

I consider these two examples of well-implemented modules.  More generally, a When in Rome™ philosophy would tell you that the way Ruby's standard library uses modules _probably_ reflects the way they were intended to be used.

In your case, I could see you implementing `NTree` this way.  The including class would have to define its own `children` method and everything in the `NTree` module would be implemented in terms of that.  It's less clear to me what the equivalent for a `DoublyLinkedList` would be in this case.

Regardless, I wouldn't recommend that approach here.   I was only describing how you might achieve the same sort of interface / interaction that `Enumerable`, `Comparable`, and other "abstract type" modules implement.  Here, I'd do what I originally suggested and simply have internal instance of a tree and a doubly linked list inside whatever class needs them.